### ISSUE TITLE
Problem combining Pymc3 and Theano function(model testing on training set instead of test set)

### ISSUE THREAD

**Issue by gbengaoti** [pymc-devs/pymc#2176]
I am trying to combine Theano functions with Pymc3 to build a recurrent neural network. However,  when I try to run the model on the test data; I notice that although the shared variable is updated, the model is tested on training data.

This `is` the code I have, I have no idea why it does not work.
```
`# CREATE PYMC3 + THEANO IMPLEMENTATION OF A SIMPLE RECURRENT NETWORK
import theano
import theano.tensor as T
import numpy as np
import pymc3 as pm
from scipy.stats import mode
theano.config.compute_test_value = 'ignore'
```
 
```
input_dim = 2
output_dim = 2
### PARAMETERS OF THE MODEL ###
hidden_dim = 64
learning_rate = 0.1
nb_epochs = 10
 
np.random.seed(0)
 
# Initialization /placeholder values
X = T.dtensor3('X')
Y = T.dtensor3('Y')
  
# generate data 
input_var = theano.shared(np.asarray(train_input).astype(np.float64))
target_var = theano.shared(np.asarray(train_output).astype(np.float64))
 
# Reference
# From paper :IMPROVING PERFORMANCE OF RECURRENT NEURAL NETWORK WITH RELU NONLINEARITY
def norm_positive_definite(r):
    A = np.dot(r, r.transpose())/hidden_dim
    values, vectors = np.linalg.eig(A)
    e = np.amax(values)
    Whh = A/e
    return Whh
 
def norm_weight(n_in, n_out):
    low = 4 *-np.sqrt(6.0/(n_in + n_out))
    high = 4 *np.sqrt(6.0/(n_in + n_out))
    W = np.random.uniform(low=low, high=high, size=(n_in,n_out))
    return W
# weights of layer
# initialize weights with random values
init_1 = norm_weight(input_dim, hidden_dim)
init_2 = norm_weight(hidden_dim, output_dim)
init_h = norm_positive_definite(np.random.normal
    (0, 1, (hidden_dim, hidden_dim)))
 
 
W1 = theano.shared(init_1.astype(theano.config.floatX))
W2 = theano.shared(init_2.astype(theano.config.floatX))
Wh = theano.shared(init_h.astype(theano.config.floatX))
 
# layer 1 which creates recurrence
L1 = theano.shared(np.zeros((8, hidden_dim)).astype(theano.config.floatX))
 
# define step
def step(v, l_tm1, W1, W2, Wh):
    layer1 = T.nnet.relu(T.dot(v, W1) + T.dot(l_tm1, Wh))
    output_layer = T.nnet.sigmoid(T.dot(layer1, W2))
    return [layer1, output_layer]
 
 
# Computational graph
[layer1s, predictions], updates = theano.scan(fn=step,
                                        sequences=dict(input=X,taps=[0]),
                                        outputs_info=[L1, None],  # same format as step function
                                        non_sequences=[W1, W2, Wh])
 
def build_rnn():
    print("Building network ...")
 
    W1=pm.Normal('w_in_1', 0, sd=1, 
                             shape=(input_dim, hidden_dim), 
                             testval=init_1.astype(np.float64))
    # Weights from 1st to 2nd layer
    W2=pm.Normal('w_2_out', 0, sd=1, 
                            shape=(hidden_dim, output_dim), 
                            testval=init_2.astype(np.float64))
     
    # Weights from hidden layer to output
    Wh=pm.Normal('w_1_2', 0, sd=1, 
                              shape=(hidden_dim,hidden_dim), 
                              testval=init_h.astype(np.float64))
 
    prediction = theano.function(inputs = [X], outputs = predictions)
     
 
    return prediction
 
with pm.Model() as rnn:
    prediction = build_rnn()
    pres = prediction(input_var.get_value())
    out = pm.Bernoulli('out',pres,observed=target_var)
     
 
with rnn:
    #Run ADVI which returns posterior means, standard deviations, 
    #and the evidence lower bound (ELBO)
    v_params = pm.variational.advi(n=50000)
    trace = pm.variational.sample_vp(v_params, draws=5000)
 
# Replace shared variables with testing set
input_var.set_value(np.asarray(test_input).astype(np.float64))
target_var.set_value(np.asarray(test_output).astype(np.float64))
 
with rnn:
    # Creater posterior predictive samples
    ppc = pm.sample_ppc(trace, samples=500)
 
# Use probability of > 0.5 to assume prediction of class 1
 
pred = ppc['out'].mean(axis=0) > 0.5
            
print (pred.shape)
 
print('Accuracy = {}%'.format((test_output == pred).mean() * 100))
 
stop = timeit.default_timer()
 
print ('It took', stop-start, 'secs')` 
```
Any ideas for this will be appreciated

**Comment by twiecki** [comment]
`pres = prediction(input_var.get_value())` this should just be `input_var`otherwise you pass in a copy of current input_var value.

**Comment by gbengaoti** [comment]
That sounds right. But I can't pass a shared variable to a function directly, because Theano expects a numeric array, not a variable.

**Comment by twiecki** [comment]
That's more of a theano question, not sure how to handle that.

**Comment by twiecki** [comment]
Feel free to keep asking questions. But I recommend the theano mailing list.

**Comment by gbengaoti** [comment]
Thank you

**Comment by twiecki** [comment]
Do let me know if you fix it, this is a great example case.

**Comment by twiecki** [comment]
@gbengaoti Did you solve the problem?

**Comment by gbengaoti** [comment]
No, not yet

**Comment by gbengaoti** [comment]
I have found that this is no longer a Theano problem. Testing the folowing code, I realized that the shared variables are being updated as expected. But it seems to get lost in PyMC3, maybe I do not understand fully, how PyMC3 tests on the holdout set, at least not enough to fix this problem.
```
`# CREATE PYMC3 + THEANO IMPLEMENTATION OF A SIMPLE RECURRENT NETWORK
import timeit
start = timeit.default_timer()
import theano
import theano.tensor as T
import numpy as np
 
theano.config.on_unused_input = 'ignore'
 
input_dim = 2
output_dim = 2
### PARAMETERS OF THE MODEL ###
hidden_dim = 64
learning_rate = 0.1
nb_epochs = 10
 
np.random.seed(0)
 
# Initialization /placeholder values
X = T.dtensor3('X')
Y = T.dtensor3('Y')
 
#.....
 
input_var = theano.shared(np.asarray(train_input).astype(np.float64))
target_var = theano.shared(np.asarray(train_output).astype(np.float64))
 
#.... 
W1 = theano.shared(init_1.astype(theano.config.floatX))
W2 = theano.shared(init_2.astype(theano.config.floatX))
Wh = theano.shared(init_h.astype(theano.config.floatX))
 
# layer 1 which creates recurrence
L1 = theano.shared(np.zeros((8, hidden_dim)).astype(theano.config.floatX))
 
# define step
def step(v, l_tm1, W1, W2, Wh):
    layer1 = T.nnet.relu(T.dot(v, W1) + T.dot(l_tm1, Wh))
    output_layer = T.nnet.sigmoid(T.dot(layer1, W2))
    return [layer1, output_layer]
 
 
# Computational graph
[layer1s, predictions], updates = theano.scan(fn=step,
                                        sequences=dict(input=X,taps=[0]),
                                        outputs_info=[L1, None],  # same format as step function
                                        non_sequences=[W1, W2, Wh])
 
 
 
prediction = theano.function(inputs=[], outputs=predictions,
                        givens=[(X,input_var),(Y,target_var)])
 
pres = prediction()
 
print (pres.shape) # should be (1500,..) -correct
     
 
# Replace shared variables with testing set
input_var.set_value(np.asarray(test_input).astype(np.float64))
target_var.set_value(np.asarray(test_output).astype(np.float64))
 
pres = prediction()
 
print (pres.shape) # should be (500,..) - correct`
```

**Comment by twiecki** [comment]
Seems like you only need to call `sample_ppc`, no?

**Comment by gbengaoti** [comment]
That's what I do here in this code but it does not work. It still does the testing on the training set.
```
`# CREATE PYMC3 + THEANO IMPLEMENTATION OF A SIMPLE RECURRENT NETWORK
import timeit
start = timeit.default_timer()
import theano
import theano.tensor as T
import numpy as np
import pymc3 as pm
from scipy.stats import mode
 
theano.config.on_unused_input = 'ignore'
theano.config.compute_test_value = 'ignore'
 
input_dim = 2
output_dim = 2
### PARAMETERS OF THE MODEL ###
hidden_dim = 64
learning_rate = 0.1
nb_epochs = 10
 
np.random.seed(0)
 
# Initialization /placeholder values
X = T.dtensor3('X')
Y = T.dtensor3('Y')
 
# .....
 
input_var = theano.shared(np.asarray(train_input).astype(np.float64))
target_var = theano.shared(np.asarray(train_output).astype(np.float64))
 
# ....
W1 = theano.shared(init_1.astype(theano.config.floatX))
W2 = theano.shared(init_2.astype(theano.config.floatX))
Wh = theano.shared(init_h.astype(theano.config.floatX))
 
# layer 1 which creates recurrence
L1 = theano.shared(np.zeros((8, hidden_dim)).astype(theano.config.floatX))
 
# define step
def step(v, l_tm1, W1, W2, Wh):
    layer1 = T.nnet.relu(T.dot(v, W1) + T.dot(l_tm1, Wh))
    output_layer = T.nnet.sigmoid(T.dot(layer1, W2))
    return [layer1, output_layer]
 
 
# Computational graph
[layer1s, predictions], updates = theano.scan(fn=step,
                                        sequences=dict(input=X,taps=[0]),
                                        outputs_info=[L1, None],  # same format as step function
                                        non_sequences=[W1, W2, Wh])
 
with pm.Model() as rnn:
    print("Building network ...")
 
    W1=pm.Normal('w_in_1', 0, sd=1, 
                             shape=(input_dim, hidden_dim), 
                             testval=init_1.astype(np.float64))
     
    W2=pm.Normal('w_2_out', 0, sd=1, 
                            shape=(hidden_dim, output_dim), 
                            testval=init_2.astype(np.float64))
     
     
    Wh=pm.Normal('w_1_2', 0, sd=1, 
                              shape=(hidden_dim,hidden_dim), 
                              testval=init_h.astype(np.float64))
 
    prediction = theano.function(inputs=[], outputs=predictions,
                            givens=[(X,input_var),(Y,target_var)])
 
    pres = prediction()
    out = pm.Bernoulli('out',pres,observed=target_var)
     
 
with rnn:
    #Run ADVI which returns posterior means, standard deviations, 
    #and the evidence lower bound (ELBO)
    v_params = pm.variational.advi(n=50000)
    trace = pm.variational.sample_vp(v_params, draws=5000)
 
# Replace shared variables with testing set
input_var.set_value(np.asarray(test_input).astype(np.float64))
target_var.set_value(np.asarray(test_output).astype(np.float64))
 
with rnn:
    # Creater posterior predictive samples
      ppc = pm.sample_ppc(trace, samples=500)
    # Use probability of > 0.5 to assume prediction of class 1
    pred = ppc['out'].mean(axis=0) > 0.5
            
print (pred.shape) #--- prints (1500,...) instead of (500,...) 
print('Accuracy = {}%'.format((test_output == pred).mean() * 100))`
```

**Comment by twiecki** [comment]
Hm, that is pretty odd, I don't see what's wrong there.

**Comment by gbengaoti** [comment]
same here. I could not find any example of combining pymc3 with Theano for deep learning, do you have any?

**Comment by gbengaoti** [comment]
I found a solution that works, I mainly separate everything into test and train functions.Here is the code:
```
`# CREATE PYMC3 + THEANO IMPLEMENTATION OF A SIMPLE RECURRENT NETWORK
import timeit
start = timeit.default_timer()
import theano
import theano.tensor as T
import numpy as np
import pymc3 as pm
from scipy.stats import mode

theano.config.on_unused_input = 'ignore'
theano.config.compute_test_value = 'ignore'

input_dim = 2
output_dim = 2
### PARAMETERS OF THE MODEL ###
hidden_dim = 64
learning_rate = 0.1
nb_epochs = 10

np.random.seed(0)

# Initialization /placeholder values
X = T.dtensor3('X')
Y = T.dtensor3('Y')

# begin by generating dataset so we have an array of lists
int2binary = {}
binary_dim = 8

largest_number = pow(2, binary_dim)
binary = np.unpackbits(np.array([range(largest_number)], dtype=np.uint8).T, axis=1)
for i in range(largest_number):
    int2binary[i] = binary[i]

X_data = []
y_data = []
dataset = np.genfromtxt('data.txt', dtype=int)


# generate data
for i in range(2000):

    a_int = dataset[i][0]
    b_int = dataset[i][1]
    c_int = dataset[i][2]


    a = int2binary[a_int]  # binary encoding

    b = int2binary[b_int]  # binary encoding

    c = int2binary[c_int]


    temp_x = []
    temp_y = []

    for position in range(binary_dim):
        X_t = np.array([a[binary_dim - position - 1], b[binary_dim - position - 1]])
        y_t = np.array([c[binary_dim - position - 1]]).T
        temp_x.append(np.asarray(X_t))

        if y_t == 0:
            temp_y.append(np.asarray([1,0]))
        else:
            temp_y.append(np.asarray([0, 1]))
        
    X_data.append(np.asarray(temp_x))
    y_data.append(np.asarray(temp_y))

NUM_EXAMPLES = 1500

test_input = X_data[NUM_EXAMPLES:]
test_output = y_data[NUM_EXAMPLES:] 

train_input = X_data[:NUM_EXAMPLES]
train_output = y_data[:NUM_EXAMPLES]

train_input_var = theano.shared(np.asarray(train_input).astype(np.float64))
train_target_var = theano.shared(np.asarray(train_output).astype(np.float64))

test_input_var = theano.shared(np.asarray(test_input).astype(np.float64))
test_target_var = theano.shared(np.asarray(test_output).astype(np.float64))

# Reference
# From paper :IMPROVING PERFORMANCE OF RECURRENT NEURAL NETWORK WITH RELU NONLINEARITY
def norm_positive_definite(r):
    A = np.dot(r, r.transpose())/hidden_dim
    values, vectors = np.linalg.eig(A)
    e = np.amax(values)
    Whh = A/e
    return Whh

def norm_weight(n_in, n_out):
    low = 4 *-np.sqrt(6.0/(n_in + n_out))
    high = 4 *np.sqrt(6.0/(n_in + n_out))
    W = np.random.uniform(low=low, high=high, size=(n_in,n_out))
    return W

# initialize weights with random values
init_1 = norm_weight(input_dim, hidden_dim)
init_2 = norm_weight(hidden_dim, output_dim)
init_h = norm_positive_definite(np.random.normal(0, 1, (hidden_dim, hidden_dim)))


W1 = theano.shared(init_1.astype(theano.config.floatX))
W2 = theano.shared(init_2.astype(theano.config.floatX))
Wh = theano.shared(init_h.astype(theano.config.floatX))

# layer 1 which creates recurrence
L1 = theano.shared(np.zeros((8, hidden_dim)).astype(theano.config.floatX))

# define step
def step(v, l_tm1, W1, W2, Wh):
    layer1 = T.nnet.relu(T.dot(v, W1) + T.dot(l_tm1, Wh))
    output_layer = T.nnet.sigmoid(T.dot(layer1, W2))
    return [layer1, output_layer]


# Computational graph
[layer1s, predictions], updates = theano.scan(fn=step,
                                        sequences=dict(input=X,taps=[0]),
                                        outputs_info=[L1, None],  # same format as step function
                                        non_sequences=[W1, W2, Wh])

with pm.Model() as rnn:
    print("Building network ...")

    W1=pm.Normal('w_in_1', 0, sd=1, 
                             shape=(input_dim, hidden_dim), 
                             testval=init_1.astype(np.float64))
    
    W2=pm.Normal('w_2_out', 0, sd=1, 
                            shape=(hidden_dim, output_dim), 
                            testval=init_2.astype(np.float64))
    
    
    Wh=pm.Normal('w_1_2', 0, sd=1, 
                              shape=(hidden_dim,hidden_dim), 
                              testval=init_h.astype(np.float64))

    train_prediction = theano.function(inputs=[], outputs=predictions,
                            givens=[(X,train_input_var),(Y,train_target_var)])

    train_pres = train_prediction()
    out = pm.Bernoulli('out',train_pres,observed=train_target_var)

    test_prediction = theano.function(inputs=[], outputs=predictions,
                            givens=[(X,test_input_var),(Y,test_target_var)])

        

with rnn:
    #Run ADVI which returns posterior means, standard deviations, 
    #and the evidence lower bound (ELBO)
    v_params = pm.variational.advi(n=50000)
    trace = pm.variational.sample_vp(v_params, draws=5000)



with rnn:
    # Creater posterior predictive samples
    # Replace shared variables with testing set
    test_pres = test_prediction()
    out_test = pm.Bernoulli('out_test',test_pres,observed=test_target_var)
    ppc = pm.sample_ppc(trace, samples=500)

# Use probability of > 0.5 to assume prediction of class 1

    pred = ppc['out_test'].mean(axis=0) > 0.5
           
print (pred.shape)

print('Accuracy = {}%'.format((test_output == pred).mean() * 100))

stop = timeit.default_timer()

print ('It took', stop-start, 'secs')`
```
